File: XCTestBootstrap/Strategies/FBManagedTestRunStrategy.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBManagedTestRunStrategy.h"

#import "FBTestManagerAPIMediator.h"
#import "FBTestManagerContext.h"
#import "FBTestRunnerConfiguration.h"
#import "FBXCTestReporter.h"
#import "XCTestBootstrapError.h"
#import "XCTestBootstrapFrameworkLoader.h"

@implementation FBManagedTestRunStrategy

#pragma mark Initializers

+ (FBFuture<NSNull *> *)runToCompletionWithTarget:(id<FBiOSTarget, FBXCTestExtendedCommands>)target configuration:(FBTestLaunchConfiguration *)configuration codesign:(nullable FBCodesignProvider *)codesign workingDirectory:(NSString *)workingDirectory reporter:(id<FBXCTestReporter>)reporter logger:(id<FBControlCoreLogger>)logger
{
  NSParameterAssert(target);
  NSParameterAssert(configuration.applicationLaunchConfiguration);
  NSParameterAssert(configuration.testBundle.path);

  NSError *error = nil;
  if (![XCTestBootstrapFrameworkLoader.allDependentFrameworks loadPrivateFrameworks:target.logger error:&error]) {
    return [XCTestBootstrapError failFutureWithError:error];
  }

  FBApplicationLaunchConfiguration *applicationLaunchConfiguration = configuration.applicationLaunchConfiguration;
  return [[FBTestRunnerConfiguration
    prepareConfigurationWithTarget:target testLaunchConfiguration:configuration workingDirectory:workingDirectory codesign:codesign]
    onQueue:target.workQueue fmap:^(FBTestRunnerConfiguration *runnerConfiguration) {
      // The launch configuration for the test bundle host.
      FBApplicationLaunchConfiguration *testHostLaunchConfiguration = [self
        prepareApplicationLaunchConfiguration:applicationLaunchConfiguration
        withTestRunnerConfiguration:runnerConfiguration];

      // Make the Context for the Test Manager.
      FBTestManagerContext *context = [[FBTestManagerContext alloc]
        initWithSessionIdentifier:runnerConfiguration.sessionIdentifier
        timeout:configuration.timeout
        testHostLaunchConfiguration:testHostLaunchConfiguration
        testedApplicationAdditionalEnvironment:runnerConfiguration.testedApplicationAdditionalEnvironment
        testConfiguration:runnerConfiguration.testConfiguration];

      // Construct and run the mediator, the core of the test execution.
      return [FBTestManagerAPIMediator
        connectAndRunUntilCompletionWithContext:context
        target:target
        reporter:reporter
        logger:logger];
    }];
}

+ (FBApplicationLaunchConfiguration *)prepareApplicationLaunchConfiguration:(FBApplicationLaunchConfiguration *)applicationLaunchConfiguration withTestRunnerConfiguration:(FBTestRunnerConfiguration *)testRunnerConfiguration
{
  return [[FBApplicationLaunchConfiguration alloc]
    initWithBundleID:testRunnerConfiguration.testRunner.identifier
    bundleName:testRunnerConfiguration.testRunner.identifier
    arguments:[self argumentsFromConfiguration:testRunnerConfiguration attributes:applicationLaunchConfiguration.arguments]
    environment:[self environmentFromConfiguration:testRunnerConfiguration environment:applicationLaunchConfiguration.environment]
    waitForDebugger:applicationLaunchConfiguration.waitForDebugger
    io:applicationLaunchConfiguration.io
    launchMode:FBApplicationLaunchModeFailIfRunning];
}

+ (NSArray<NSString *> *)argumentsFromConfiguration:(FBTestRunnerConfiguration *)configuration attributes:(NSArray<NSString *> *)attributes
{
  return [(configuration.launchArguments ?: @[]) arrayByAddingObjectsFromArray:(attributes ?: @[])];
}

+ (NSDictionary<NSString *, NSString *> *)environmentFromConfiguration:(FBTestRunnerConfiguration *)configuration environment:(NSDictionary<NSString *, NSString *> *)environment
{
  NSMutableDictionary<NSString *, NSString *> *mEnvironment = (configuration.launchEnvironment ?: @{}).mutableCopy;
  if (environment) {
    [mEnvironment addEntriesFromDictionary:environment];
  }
  return [mEnvironment copy];
}

@end

File: XCTestBootstrap/Strategies/FBManagedTestRunStrategy.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import FBControlCore
import Foundation

enum ManagedTestRunError: Error, LocalizedError {
  case frameworkLoadingFailed(underlying: Error)

  public var errorDescription: String? {
    switch self {
    case let .frameworkLoadingFailed(underlying):
      return underlying.localizedDescription
    }
  }
}

public final class FBManagedTestRunStrategy {

  public static func runToCompletion(withTarget target: any XCTestExtendedTarget, configuration: TestLaunchConfiguration, codesign: FBCodesignProvider?, workingDirectory: String, reporter: FBXCTestReporter, logger: FBControlCoreLogger) async throws {
    do {
      try XCTestBootstrapFrameworkLoader.allDependentFrameworks.loadPrivateFrameworks(target.logger)
    } catch {
      throw ManagedTestRunError.frameworkLoadingFailed(underlying: error)
    }

    let runnerConfiguration = try await TestRunnerConfiguration.prepareConfiguration(
      withTarget: target,
      testLaunchConfiguration: configuration,
      workingDirectory: workingDirectory,
      codesign: codesign
    )

    let testHostLaunchConfiguration = prepareApplicationLaunchConfiguration(configuration.applicationLaunchConfiguration, withTestRunnerConfiguration: runnerConfiguration)

    let context = TestManagerContext(
      sessionIdentifier: runnerConfiguration.sessionIdentifier,
      timeout: configuration.timeout,
      testHostLaunchConfiguration: testHostLaunchConfiguration,
      testedApplicationAdditionalEnvironment: runnerConfiguration.testedApplicationAdditionalEnvironment,
      testConfiguration: runnerConfiguration.testConfiguration
    )

    try await TestManagerAPIMediator.connectAndRunUntilCompletion(
      with: context,
      target: target,
      reporter: reporter,
      logger: logger
    )
  }

  private static func prepareApplicationLaunchConfiguration(_ applicationLaunchConfiguration: FBApplicationLaunchConfiguration, withTestRunnerConfiguration testRunnerConfiguration: TestRunnerConfiguration) -> FBApplicationLaunchConfiguration {
    FBApplicationLaunchConfiguration(
      bundleID: testRunnerConfiguration.testRunner.identifier,
      bundleName: testRunnerConfiguration.testRunner.identifier,
      arguments: arguments(fromConfiguration: testRunnerConfiguration, attributes: applicationLaunchConfiguration.arguments),
      environment: environment(fromConfiguration: testRunnerConfiguration, environment: applicationLaunchConfiguration.environment),
      waitForDebugger: applicationLaunchConfiguration.waitForDebugger,
      io: applicationLaunchConfiguration.io,
      launchMode: .relaunchIfRunning
    )
  }

  private static func arguments(fromConfiguration configuration: TestRunnerConfiguration, attributes: [String]) -> [String] {
    configuration.launchArguments + attributes
  }

  private static func environment(fromConfiguration configuration: TestRunnerConfiguration, environment: [String: String]) -> [String: String] {
    var mEnvironment = configuration.launchEnvironment
    for (key, value) in environment {
      mEnvironment[key] = value
    }
    return mEnvironment
  }
}
